test/provider_os_memory.cpp

```diff
@@ -39,396 +47,243 @@ static const char *Native_error_str[] = {
     "force purging failed", // UMF_OS_RESULT_ERROR_PURGE_FORCE_FAILED
 };
 
+// test helpers
+
 static int compare_native_error_str(const char *message, int error) {
     const char *error_str = Native_error_str[error - UMF_OS_RESULT_SUCCESS];
     size_t len = strlen(error_str);
     return strncmp(message, error_str, len);
 }
 
-TEST_F(test, provider_os_memory_create_destroy) {
-    umf_result_t umf_result;
-    umf_memory_provider_handle_t os_memory_provider = nullptr;
-    umf_os_memory_provider_params_t os_memory_provider_params =
-        UMF_OS_MEMORY_PROVIDER_PARAMS_TEST;
+using providerCreateExtParams = std::tuple<umf_memory_provider_ops_t *, void *>;
 
-    umf_result = umfMemoryProviderCreate(&UMF_OS_MEMORY_PROVIDER_OPS,
-                                         &os_memory_provider_params,
-                                         &os_memory_provider);
-    ASSERT_EQ(umf_result, UMF_RESULT_SUCCESS);
-    ASSERT_NE(os_memory_provider, nullptr);
-
-    umfMemoryProviderDestroy(os_memory_provider);
-}
+umf::provider_unique_handle_t
+providerCreateExt(providerCreateExtParams params) {
+    umf_memory_provider_handle_t hProvider;
+    auto [provider_ops, provider_params] = params;
 
-TEST_F(test, provider_os_memory_create_UMF_NUMA_MODE_NOT_SUPPORTED) {
-    umf_result_t umf_result;
-    umf_memory_provider_handle_t os_memory_provider = nullptr;
-    umf_os_memory_provider_params_t os_memory_provider_params =
-        UMF_OS_MEMORY_PROVIDER_PARAMS_TEST;
-
-    // NUMA binding mode not supported for UMF_VISIBILITY_SHARED
-    os_memory_provider_params.visibility = UMF_VISIBILITY_SHARED;
-    os_memory_provider_params.numa_mode = UMF_NUMA_MODE_BIND;
-
-    umf_result = umfMemoryProviderCreate(&UMF_OS_MEMORY_PROVIDER_OPS,
-                                         &os_memory_provider_params,
-                                         &os_memory_provider);
-    ASSERT_EQ(umf_result, UMF_RESULT_ERROR_INVALID_ARGUMENT);
-    ASSERT_EQ(os_memory_provider, nullptr);
-}
-
-TEST_F(test, provider_os_memory_create_MBIND_FAILED) {
-    umf_result_t umf_result;
-    umf_memory_provider_handle_t os_memory_provider = nullptr;
-    umf_os_memory_provider_params_t os_memory_provider_params =
-        UMF_OS_MEMORY_PROVIDER_PARAMS_TEST;
+    auto ret =
+        umfMemoryProviderCreate(provider_ops, provider_params, &hProvider);
+    EXPECT_EQ(ret, UMF_RESULT_SUCCESS);
+    EXPECT_NE(hProvider, nullptr);
 
-    // wrong NUMA flags
-    os_memory_provider_params.numa_flags = (unsigned int)-1;
+    // capture provider and destroy it
+    auto providerDestructor = [](umf_memory_provider_handle_t hProvider) {
```

Comment:
nit: you don't need to create a custom destructor here, you can just pass `umfMemoryProviderDestroy` directly to unique_ptr ctor, like this: `return umf::provider_unique_handle_t(hProvider, &umfMemoryProviderDestroy);`

This was needed for poolCreateExt because we wanted to destroy both pool and provider at the same time.

Reply:
Done